src/emqx_mod_topic_metrics.erl

```diff
@@ -364,16 +378,36 @@ number_of_registered_topics() ->
     proplists:get_value(size, ets:info(?TAB)).
 
 calculate_speed(CurVal, #speed{last = Last,
-                               last_v = LastVal
-                              }) ->
+    last_v = LastVal
+}) ->
     %% calculate the current speed based on the last value of the counter
     CurSpeed = (CurVal - LastVal) / ?TICKING_INTERVAL,
-    #speed{last = mma(?SPEED_AVERAGE_WINDOW_SIZE, Last, CurSpeed),
-           last_v = CurVal,
-           last_medium = mma(?SPEED_MEDIUM_WINDOW_SIZE, Last, CurSpeed),
-           last_long = mma(?SPEED_LONG_WINDOW_SIZE, Last, CurSpeed)
-          }.
+    #speed{
+        last_v = CurVal,
+        last = last_short(Last, CurSpeed),
+        last_medium = last_medium(Last, CurSpeed),
+        last_long = last_long(Last, CurSpeed)
+    }.
 
 %% Modified Moving Average ref: https://en.wikipedia.org/wiki/Moving_average
 mma(WindowSize, LastSpeed, CurSpeed) ->
-  (LastSpeed * (WindowSize - 1) + CurSpeed) / WindowSize.
+    (LastSpeed * (WindowSize - 1) + CurSpeed) / WindowSize.
+
+last_short(LastSpeed, CurSpeed) ->
+    mma(?SPEED_AVERAGE_WINDOW_SIZE, LastSpeed, CurSpeed).
+
+last_medium(LastSpeed, CurSpeed) ->
+    mma(?SPEED_MEDIUM_WINDOW_SIZE, LastSpeed, CurSpeed).
+
+last_long(LastSpeed, CurSpeed) ->
+    mma(?SPEED_LONG_WINDOW_SIZE, LastSpeed, CurSpeed).
+
+last_stats_all(LastSpeed, CurSpeed) ->
+    #last_stats{
```

Comment:
records defined in a module locally are not quite useful.
If a record schema is to be shared with other modules, it is usually defined in a header file (`.hrl`)

However, given that the plan is to expose the values to a very remote component (where the WEB UI is rendered),
I'd suggest to use a map instead. e.g.
```
#{short => ..., medium => ..., long => ...}
```